File: incre-tests/autolifter/dac/min.f
Config ExtraGrammar = "AutoLifter";

Inductive List = cons {Int, List} | nil Unit;

length = fix (\f: List->Int. \x: List.
  match x with
    cons {h, t} -> + (f t) 1
  | nil _ -> 0
  end
);

concat = fix (\f: List->List->List. \x: List. \y: List.
  match x with
    cons {h, t} -> cons {h, f t y}
  | nil _ -> y
  end
);

split = \xs: List. fix (\f: List->Int->{List, List}. \x: List. \n: Int.
  if (< n 1) then {nil unit, x}
  else match x with
    cons {h, t} -> let res = f t (- n 1) in {cons {h, res.1}, res.2}
  | _ -> {nil unit, nil unit}
  end
) xs (/ (length xs) 2);

dac = \v: List -> Int. \xs: List.
  let run = fix (\f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {_, nil _} -> xs
    | _ -> let sp = split xs in
      concat (f sp.1) (f sp.2)
    end
  ) in v (run xs);

inf = 100;

/*User provided programs*/

min = fix (\f: List -> Int. \xs: List.
  match xs with
    nil _ -> inf
  | cons {h, t} ->
    let res = f t in if < res h then res else h
  end
);

main = dac min;

File: ml-benchmark/autolifter/dac/min.ml
type list_ = Cons of int * list_ | Nil

let rec length x =
  match x with
  | Cons (_, t) -> 1 + length t
  | Nil -> 0

let rec concat x y =
  match x with
  | Cons (h, t) -> Cons (h, concat t y)
  | Nil -> y

let split xs =
  let rec f x n =
    if n < 1 then
      (Nil, x)
    else
      match x with
      | Cons (h, t) ->
          let res = f t (n - 1) in
          match res with
          | (l, r) -> (Cons (h, l), r)
      | _ -> (Nil, Nil)
  in
  f xs (length xs / 2)

let dac v xs =
  let rec run xs =
    match xs with
    | Nil -> xs
    | Cons (_, t) ->
        match t with
        | Nil -> xs
        | _ ->
            let sp = split xs in
            match sp with
            | (a, b) ->
                let ra = run a in
                let rb = run b in
                concat ra rb
  in
  v (run xs)

let inf = 100

let rec min xs =
  match xs with
  | Nil -> inf
  | Cons (h, t) ->
      let res = min t in
      if res < h then res else h

let program = dac min